distributed/tests/test_client.py

```diff
@@ -7500,3 +7501,76 @@ async def test_wait_for_workers_updates_info(c, s):
     async with Worker(s.address):
         await c.wait_for_workers(1)
         assert c.scheduler_info()["workers"]
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_client_story(c, s, *workers):
+    f = c.submit(inc, 1)
+    assert await f == 2
+    story = await c.story(f.key)
+
+    assert_story(
+        story,
+        [
+            (f.key, "released", "waiting", {f.key: "processing"}),
+            (f.key, "waiting", "processing", {}),
+            (f.key, "processing", "memory", {}),
+            (f.key, "compute-task"),
+            (f.key, "released", "waiting", "waiting", {f.key: "ready"}),
+            (f.key, "waiting", "ready", "ready", {f.key: "executing"}),
+            (f.key, "ready", "executing", "executing", {}),
+            (f.key, "put-in-memory"),
+            (f.key, "executing", "memory", "memory", {}),
+        ],
+        ordered_timestamps=False,
+    )
+
+    stimulus_ids = {ev[-2] for ev in story}
+    assert len(stimulus_ids) == 3
+
+
+class WorkerBrokenStory(Worker):
+    async def get_story(self, *args, **kw):
+        raise CommClosedError
+
+
+@gen_cluster(client=True, Worker=WorkerBrokenStory)
+@pytest.mark.parametrize("on_error", ["ignore", "return", "raise"])
+async def test_client_story_failed_worker(c, s, a, b, on_error):
+    f = c.submit(inc, 1)
+    coro = c.story(f.key, on_error=on_error)
+    await f
+
+    if on_error == "raise":
+        with pytest.raises(CommClosedError) as e:
+            await coro
+    elif on_error == "ignore":
+        story = await coro
+        assert_story(
+            story,
+            [
+                (f.key, "released", "waiting", {f.key: "processing"}),
+                (f.key, "waiting", "processing", {}),
+                (f.key, "processing", "memory", {}),
+                ("worker-story-retrieval-failure",),
+                ("worker-story-retrieval-failure",),
+            ],
+        )
```

Comment:
This seems very precise.  I don't think that we need all of these details.  I think that instead you're looking to test that you get the outputs contain `s.story(f.key)` but don't contain the worker's story.  Is that correct?  If so, then I would vastly prefer a test that didn't test the exact outputs, but instead did something like the following:

```python
assert set(story).issuperset(s.story(f.key))
assert not set(story) & a.story(f.key) 
assert not set(story) & b.story(f.key) 
```

This seems to test the thing that you want to test (that we get some outputs but not all of them) in a way that is far less brittle to future changes.  

Thoughts?

Reply:
My understanding of yesterday's conversation with @fjetter was that the use of `assert_story` was acceptable at present.